### PR TITLE
Add support to materialize to write documents out to files.

### PR DESCRIPTION
**Pull request by eric-anderson**
Several ways to configure for high customizability.

### CODE REVIEW

File: lib/sycamore/sycamore/lineage.py [hunk 1]
```diff
@@ -1,38 +1,100 @@
...
-    def __init__(self, child: Node, **kwargs):
+    def __init__(self, child: Node, context: Context, path: Optional[Path | str | dict], **kwargs):
         assert isinstance(child, Node)
+
+        self._root = None
+        if path is None:
+            pass
+        elif isinstance(path, str) or isinstance(path, Path):
+            (self._fs, self._root) = self.infer_fs(str(path))
+            self._doc_to_name = self.doc_to_name
```
[the remark this answers is not in the document]
Comment (HenryL27) [reply lib/sycamore/sycamore/lineage.py:24]: oh, found it

File: lib/sycamore/sycamore/lineage.py [hunk 2]
```diff
@@ -1,38 +1,100 @@
...
+            self._clean_root = True
+        elif isinstance(path, dict):
+            self._root = Path(path["root"])
+            if "fs" in path:
+                self._fs = path["fs"]
+            else:
+                (self._fs, self._root) = self.infer_fs(str(self._root))
+            self._doc_to_name = path.get("name", None) or self.doc_to_name
```
[the remark this answers is not in the document]
Comment (HenryL27) [reply lib/sycamore/sycamore/lineage.py:32]: maybe make sure that path['name'] is callable

**Review by HenryL27 — COMMENTED** [review]
couple nits but I think it's good. Doing stuff with this output comes later, yes?

File: lib/sycamore/sycamore/lineage.py [hunk 3]
```diff
@@ -1,38 +1,100 @@
...
+
+        self._root = None
+        if path is None:
+            pass
+        elif isinstance(path, str) or isinstance(path, Path):
+            (self._fs, self._root) = self.infer_fs(str(path))
+            self._doc_to_name = self.doc_to_name
+            self._clean_root = True
+        elif isinstance(path, dict):
+            self._root = Path(path["root"])
```
[the remark this answers is not in the document]
Comment (eric-anderson) [reply lib/sycamore/sycamore/lineage.py:27]: Done.

File: lib/sycamore/sycamore/lineage.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eric-anderson) [reply lib/sycamore/sycamore/lineage.py:32]: Done.

File: lib/sycamore/sycamore/lineage.py [hunk 4]
```diff
@@ -1,38 +1,100 @@
...
+        if self._root is not None:
+            import numpy
+
+            self.cleanup()
+
+            @rename("lineage-materialize")
+            def ray_callable(ray_input: dict[str, numpy.ndarray]) -> dict[str, numpy.ndarray]:
+                for s in ray_input.get("doc", []):
+                    self.save(Document.deserialize(s))
+                return ray_input
```
[the remark this answers is not in the document]
Comment (eric-anderson) [reply lib/sycamore/sycamore/lineage.py:53]: Yes, when we continue from a materialize node we need them to be able to continue.

File: lib/sycamore/sycamore/lineage.py [hunk 5]
```diff
@@ -1,38 +1,100 @@
...
+            self._clean_root = path.get("clean", True)
+        else:
+            assert False, f"unsupported type ({type(path)}) for path argument, expected str, Path, or dict"
+
         super().__init__(child, **kwargs)
 
     def execute(self, **kwargs) -> "Dataset":
+        # right now, the only thing we can do is save data, so do it in parallel.
+        # once we support validation we won't be able to run the validation in parallel.
+        # non-shared filesystems will also eventually be a problem but we can put it off for now.
```
[the remark this answers is not in the document]
Comment (eric-anderson) [reply lib/sycamore/sycamore/lineage.py:42]: The logic that leads to retry.  Updated comment.

File: lib/sycamore/sycamore/lineage.py [hunk 6]
```diff
@@ -1,38 +1,100 @@
...
+        [d for d in docs if isinstance(d, MetadataDocument)]
+        # logging.info(f"Found {len(md)} md documents")
+        # logging.info(f"\n{pprint.pformat(md)}")
         return docs
+
+    @staticmethod
+    def infer_fs(path: str) -> "pyarrow.FileSystem":
+        from pyarrow import fs
+
+        (fs, root) = fs.FileSystem.from_uri(path)
```
[the remark this answers is not in the document]
Comment (eric-anderson) [reply lib/sycamore/sycamore/lineage.py:73]: ack

File: lib/sycamore/sycamore/lineage.py [hunk 7]
```diff
@@ -1,38 +1,100 @@
...
+        return (fs, Path(root))
+
+    def save(self, doc: Document) -> None:
+        assert self._root is not None
+        name = self._doc_to_name(doc)
+        path = self._root / name
+        with self._fs.open_output_stream(str(path)) as out:
+            out.write(doc.serialize())
```
[the remark this answers is not in the document]
Comment (eric-anderson) [reply lib/sycamore/sycamore/lineage.py:81]: We'd have to figure out how we're going to batch. I'll put that off until we have a need for it.  It's possible that the answer will be that we have a fs writer that buffers and combines.

**Review by HenryL27 — APPROVED** [review]

**Review by HenryL27 — APPROVED** [review]
lgtm; as discussed, drop df4 and add one after run tests

**Pull request merged**